### DISCUSSION TITLE
Is it possible to use left_join with group_by and subqueries without forcing it through sql_query?

### DISCUSSION THREAD

**Discussion by davidgxue** [diesel-rs/diesel#3550] [Q&A]
I am using Postgres and below is what I am trying to achieve. I initially wrote stuff with LATERAL join but it doesn't look like diesel supports Postgres specific stuff like that so I switched it to something like below. So far nothing I have tried works tho. Only had success doing everything in a string then doing sql_query() with it.
```
SELECT training.training_id, training.name, training.description, filtered_status_table.status
FROM training 
LEFT JOIN (
  SELECT ts1.training_id, ts1.status, ts1.status_updated_on
  FROM training_status ts1
  WHERE ts1.status_updated_on = (
    SELECT MAX(ts2.status_updated_on)
    FROM training_status ts2
    WHERE ts2.training_id = ts1.training_id
  )
) AS filtered_status_table 
ON training.training_id = filtered_status_table.training_id 
AND filtered_status_table.training_id = training.training_id;
```

And here are the schemas. I am able to handle the TrainingStatusType enum just fine (tested with simpler left_joins), so that's not the issue
```
diesel::table! {
    use diesel::sql_types::*;
    use super::sql_types::TrainingStatusType;

    training_status (training_id, status) {
        training_id -> Uuid,
        status -> TrainingStatusType,
        status_updated_on -> Timestamp,
        status_description -> Nullable<Text>,
    }
}
diesel::table! {
    use diesel::sql_types::*;

    training (training_id) {
        training_id -> Uuid,
        name -> Text,
        description -> Nullable<Text>,
        dataset_id -> Uuid,
        created_on -> Timestamp,
    }
}
```
I am essentially trying to get training related info + status by joining the two tables together. The subquery is needed because in the training_status table I need to select the row with the latest timestamp given a training_id (so multiple rows of same training_id exists in that table, selecting the latest one).

**The main problems I have are the aliasing of the left_joined table and the subquery (see line "WHERE ts2.training_id = ts1.training_id". For aliasing, it doesn't seem like Diesel has a way to alias a left_joined table and so I can't select rows from both columns. Maybe diesel doesn't support that yet?**

My terrible attempt that has so many compiler errors:
```
let (ts1, ts2) = alias!(training_status as ts1, training_status as ts2);

let filtered_status_table = ts2.select(dsl::max(ts2.field(training_status::status_updated_on))).
filter(ts2.field(training_status::training_id).eq(ts1.field(training_status::training_id)));

let res = training::table.select(
(training::training_id, training::name, training::description, filtered_status_table.field(training_status::status))
).left_join(filtered_status_table.on(training::training_id.eq(filtered_status_table.field(training_status::training_id))))
.load::<TrainingBasicInfo>(conn);
```

Errors such as 
```no method `filter` on type `<Alias<ts2> as SelectDsl<{unknown}>>::Output```

```no method `eq` on type `AliasedField<ts2, training_id>```

```no method `left_join` on type `<table as SelectDsl<(training_id, name, description, {unknown})>>::Output`rust```

I have made so many attempts that I lost track. What I have above is just one version..

**Sorry about the long post, new to rust and diesel, any help is much appreciated!**

**Comment by weiznich** [comment]
The issue with your query is not about `alias!` or `left_join` but rather that diesel's built-in DSL does not support joining subqueries yet. This is a known issue and an area where we look for contributions.

**Reply by davidgxue** [reply]
Thanks for the response. I think that's kinda what I was getting at as well. Make sense. 

I did have a workaround in mind that technically doesn't use subquery in this specific case, but I am also not fully sure how to get it to work. Here's some short code and error message if you/someone doesn't mind taking a look. Or perhaps this isn't supported either?

```
training::table.left_join(training_status::table.on(training_status::training_id.eq(training::training_id)))
        .group_by(training::training_id)
        .select(training::id, training::name, training::description, training_status::status); 
//then add more stuff to this if this works
```

Error
```
the trait bound `training::columns::training_id: IsContainedInGroupBy<training_status::columns::status>` is not satisfied
the following other types implement trait `IsContainedInGroupBy<T>`:
  <training::columns::training_id as IsContainedInGroupBy<training::columns::created_by_customer_id>>
  <training::columns::training_id as IsContainedInGroupBy<training::columns::created_on>>
  <training::columns::training_id as IsContainedInGroupBy<training::columns::description>>
  <training::columns::training_id as IsContainedInGroupBy<training::columns::name>>
and 2 others
```

So I think the problem is if I do group by and specify the column is from training (even tho that column is a foreign key/shared between two tables), then I cannot select stuff not in training table. Is there something I am missing here?

**Reply by weiznich** [reply]
Well this query is rejected by diesel as the select clause misses aggregate and non-aggregate expressions. Your group by clause groups by the primary key of the `training` table. That implies that all columns from this table are considered to be non-aggregate, while all columns coming from `training_status` are considered to be aggregated. You cannot mix them in your select clause. It's hard to tell whether you should solve this by :
* removing `training_status::status` from your select clause
* Add a aggregate function (like `min`/`max`/`avg`…) around the `training_status::status` field
* Add said field to the group by clause.
